File: shared_rs/src/application/pipeline_factory.rs
//! Shared application: pipeline factory and stage registry.
//!
//! Provides a global-style registry of stage builders and functions to
//! construct [`Pipeline`] instances from TOML configuration tables.

use std::collections::HashMap;

use crate::domain::pipeline::Stage;
use crate::domain::pipeline_runner::Pipeline;

/// A builder function that constructs a [`Stage`] from a config map.
pub type StageBuilder = fn(HashMap<String, String>) -> anyhow::Result<Box<dyn Stage>>;

/// Registry of stage type names to their builder functions.
#[derive(Default)]
pub struct StageRegistry {
    builders: HashMap<String, StageBuilder>,
}

impl StageRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Register a stage builder under the given type name.
    pub fn register(&mut self, type_name: impl Into<String>, builder: StageBuilder) {
        self.builders.insert(type_name.into(), builder);
    }

    /// Build a single [`Pipeline`] from a pipeline definition and stages config.
    ///
    /// `pipeline_def` must contain a `"stages"` key with a comma-separated
    /// list of stage names. Each stage name is looked up in `stages_config`
    /// to retrieve its builder parameters.
    pub fn build_pipeline(
        &self,
        name: &str,
        stage_names: &[String],
        stages_config: &HashMap<String, HashMap<String, String>>,
    ) -> anyhow::Result<Pipeline> {
        let mut stages: Vec<Box<dyn Stage>> = Vec::new();

        for stage_name in stage_names {
            let stage_cfg = stages_config
                .get(stage_name.as_str())
                .cloned()
                .unwrap_or_default();

            let type_name = stage_cfg
                .get("type")
                .cloned()
                .unwrap_or_else(|| stage_name.clone());

            let builder = self.builders.get(type_name.as_str()).ok_or_else(|| {
                anyhow::anyhow!(
                    "Unknown stage type: {type_name:?}. Registered types: {:?}",
                    self.builders.keys().collect::<Vec<_>>()
                )
            })?;

            stages.push(builder(stage_cfg)?);
        }

        Ok(Pipeline::new(name, stages))
    }
}

File: tts_python/application/pipeline_factory.py
"""
Shared application: pipeline factory and stage registry.

Provides the global registry of stage builders and functions to
construct Pipeline instances from TOML configuration dicts.
"""

from __future__ import annotations

from copy import deepcopy
from typing import Callable

from shared.domain.pipeline import Stage
from shared.domain.pipeline_runner import Pipeline


# ---------------------------------------------------------------------------
# Global registry: stage_type_name -> builder function
# ---------------------------------------------------------------------------

_STAGE_BUILDERS: dict[str, Callable[[dict], Stage]] = {}


def register_stage(type_name: str):
    """
    Decorator for registering stage builder functions.

    Usage::

        @register_stage("huggingface_asr")
        def _build_hf_asr(cfg: dict) -> Stage:
            from ptt.infra_huggingface.stage import HuggingFaceASRStage
            return HuggingFaceASRStage(**cfg)
    """

    def wrapper(fn: Callable[[dict], Stage]):
        _STAGE_BUILDERS[type_name] = fn
        return fn

    return wrapper


# ---------------------------------------------------------------------------
# Pipeline construction
# ---------------------------------------------------------------------------


def build_pipeline(
    name: str,
    pipeline_def: dict,
    stages_config: dict,
) -> Pipeline:
    """
    Construct a :class:`Pipeline` from parsed TOML config dicts.

    Parameters
    ----------
    name:
        Human-readable pipeline name (used for lookups).
    pipeline_def:
        Dict with at least a ``"stages"`` key listing stage names.
    stages_config:
        The full ``[stages]`` section from config, keyed by stage name.
    """
    stages: list[Stage] = []
    for stage_name in pipeline_def["stages"]:
        stage_cfg = deepcopy(stages_config.get(stage_name, {}))
        type_name = stage_cfg.pop("type", stage_name)
        builder = _STAGE_BUILDERS.get(type_name)
        if not builder:
            raise ValueError(
                f"Unknown stage type: {type_name!r}. "
                f"Registered types: {sorted(_STAGE_BUILDERS)}"
            )
        stages.append(builder(stage_cfg))
    return Pipeline(name=name, stages=stages)


def load_pipelines(config_raw: dict) -> dict[str, Pipeline]:
    """
    Load all named pipelines from a raw TOML config dict.

    Expects optional top-level keys ``"pipelines"`` and ``"stages"``.
    Returns an empty dict if no ``[pipelines]`` section is present.
    """
    pipelines_section = config_raw.get("pipelines", {})
    stages_section = config_raw.get("stages", {})

    if not pipelines_section:
        return {}

    pipelines: dict[str, Pipeline] = {}
    for name, pdef in pipelines_section.items():
        pipelines[name] = build_pipeline(name, pdef, stages_section)
    return pipelines


__all__ = [
    "register_stage",
    "build_pipeline",
    "load_pipelines",
]
